onvm/onvm_nflib/onvm_nflib.c

```diff
@@ -1411,3 +1525,239 @@ onvm_nflib_stats_summary_output(uint16_t id) {
+                        RTE_LOG(INFO, APP, "Could not dequeue %s, ring is empty\n", nf_name);
+                        break;
+                }
+                sem = sem_open(dequeued_nf->pool_status.pool_mutex_name, 0, 0666, 0);
+                dequeued_nf->pool_status.pool_sleep_state = 0;
+                if (sem_post(sem) < 0) {
+                        RTE_LOG(INFO, APP, "Could not post to semaphore\n");
+                        break;
+                }
+        }
+        if (refill >= 0) {
+                pool_ctx->refill = refill;
+        }
+
+        if (rte_ring_count(nf_pool_ring) < pool_ctx->refill) {
+                total_refill = pool_ctx->refill - rte_ring_count(nf_pool_ring);
+                spawned_nf_count = onvm_nflib_fork_pool_nfs(pool_ctx->nf_name, pool_ctx->args,
+                                                            nf_pool_ring, total_refill);
+                RTE_LOG(INFO, APP, "Refilled NF pool with %d NF's\n", spawned_nf_count);
+        }
+
+        return i;
+}
+
+struct onvm_nf *
+onvm_nflib_fork(const char *nf_name, const char *nf_args) {
+        char *go_script_path;
+        int instance_id;
+
+        if (nf_name == NULL) {
+                RTE_LOG(INFO, APP, "onvm_nflib_fork(): nf_name is null\n");
+                return NULL;
+        }
+        if (nf_args == NULL) {
+                RTE_LOG(INFO, APP, "onvm_nflib_fork(): nf_args is null\n");
+                return NULL;
+        }
+
+        instance_id = onvm_nflib_request_next_instance_id();
+        go_script_path = onvm_nflib_get_go_script_path();
+        if (go_script_path == NULL) {
+                return NULL;
+        }
+        if (fork() == 0) {
+                char *command;
+                int ret;
+                ret = asprintf(&command, "%s %s %s", go_script_path, nf_name, nf_args);
+                if (ret < 0) {
+                        RTE_LOG(INFO, APP, "Could not allocate command string for bash script call\n");
+                        return NULL;
+                }
+                ret = system(command);
+                if (ret < 0) {
+                        RTE_LOG(INFO, APP, "Could not execute NF go script\n");
+                        return NULL;
+                }
+        }
+
+        return &nfs[instance_id];
+}
+
+char *
+onvm_nflib_get_go_script_path(void) {
+        char *token;
+        const char *examples_path = "/examples/";
+        const char *go_script = "start_nf.sh";
+        char *go_script_path;
+        uint16_t total_directory_size, wd_nf_len, token_len, cwd_len, go_script_len;
+        char *cwd = rte_malloc(0, sizeof(char) * 4096, 0);
+
+        if (getcwd(cwd, 4096) == NULL) {
+                RTE_LOG(INFO, APP, "Could not extract current working CWD");
+                return NULL;
+        }
+
+        token = strstr(cwd, examples_path);
+        *token = '\0';
+        token_len = strlen(token);
+        cwd = cwd + token_len;
+
+        cwd_len = strlen(cwd);
```

Comment:
I got a compile error on this specific line. 

In file included from /usr/include/string.h:495,                                                                                                              
                 from /local/onvm/openNetVM/onvm/onvm_nflib/onvm_includes.h:62,                                                                               
                 from /local/onvm/openNetVM/onvm/onvm_nflib/onvm_nflib.c:56:                                                                                  
In function ‘strncat’,                                                                                                                                        
    inlined from ‘onvm_nflib_get_go_script_path’ at /local/onvm/openNetVM/onvm/onvm_nflib/onvm_nflib.c:1721:9:                                                
/usr/include/x86_64-linux-gnu/bits/string_fortified.h:136:10: error: ‘__builtin_strncat’ output truncated before terminating nul copying as many bytes from a 
string as its length [-Werror=stringop-truncation]
  136 |   return __builtin___strncat_chk (__dest, __src, __len, __bos (__dest));
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/local/onvm/openNetVM/onvm/onvm_nflib/onvm_nflib.c: In function ‘onvm_nflib_get_go_script_path’:                                                              
/local/onvm/openNetVM/onvm/onvm_nflib/onvm_nflib.c:1715:19: note: length computed here
 1715 |         cwd_len = strlen(cwd);
      |                   ^~~~~~~~~~~
cc1: all warnings being treated as errors   
![image](https://user-images.githubusercontent.com/23443297/96041873-16339400-0e9f-11eb-9815-caf02152c989.png)

Reply:
Could you tell me what gcc version this is from? Maybe it's b/c I ran this on an older version of ubuntu.

Reply:
gcc (Ubuntu 9.3.0-10ubuntu2) 9.3.0
I think you're right, I'm testing on Ubuntu 20. Maybe I can try this on Ubuntu 18.

Reply:
if we are going to merge this it will need to run on ubuntu20's super strict gcc. @sreyanalla  fixed a lot of these errors when porting ONVM so you can check the commit log from that PR if you want to see how to get around it.

Reply:
@WilliamMaa I just tried compiling and running an example, and it worked fine on Ubuntu 20 with gcc 9.3.0. Is there anything special you did to compile? I didn't get any weird compiler warnings or errors.

Reply:
@kevindweb I'll try again, I don't think I did anything special.